src/hamiltonian.jl

```diff
@@ -0,0 +1,61 @@
+# TODO: add a type for kinetic energy
+
+struct Hamiltonian{M<:AbstractMetric}
+    metric      ::  M
+    _logπ       ::  Function
+    _dlogπdθ    ::  Function
+end
+
+# TODO: implement a helper function for those only _logπ is provided and use AD to provide _dlogπdθ
+
+function _dHdθ(h::Hamiltonian, θ::AbstractVector{T}) where {T<:Real}
+    return -h._dlogπdθ(θ)
+end
+
+function _dHdr(h::Hamiltonian{UnitMetric}, r::AbstractVector{T}) where {T<:Real}
+    return r
+end
+
+function _dHdr(h::Hamiltonian{DiagMetric{T}}, r::AbstractVector{T}) where {T<:Real}
+    return h.metric.M⁻¹ .* r
+end
+
+function _dHdr(h::Hamiltonian{DenseMetric{T}}, r::AbstractVector{T}) where {T<:Real}
+    return h.metric.M⁻¹ * r
+end
+
+function _H(h::Hamiltonian, θ::AbstractVector{T}, r::AbstractVector{T}) where {T<:Real}
+    return _K(h, r, θ) + _V(h, θ)
+end
+
+function _V(h::Hamiltonian, θ::AbstractVector{T}) where {T<:Real}
+    return -h._logπ(θ)
+end
+
+# Kinetic energy
+# NOTE: the general form of K depends on both θ and r
+function _K(h::Hamiltonian{UnitMetric}, r::AbstractVector{T}, ::AbstractVector{T}) where {T<:Real}
+    return sum(abs2, r) / 2
+end
+
+function _K(h::Hamiltonian{DiagMetric{T}}, r::AbstractVector{T}, ::AbstractVector{T}) where {T<:Real}
+    return sum(abs2, r .* sqrt.(h.metric.M⁻¹)) / 2
+end
+
+function _K(h::Hamiltonian{DenseMetric{T}}, r::AbstractVector{T}, ::AbstractVector{T}) where {T<:Real}
+    return p' * h.metric.M⁻¹ * p / 2
+end
+
+# Momentum sampler
+function rand_momentum(h::Hamiltonian{UnitMetric}, θ::AbstractVector{T}) where {T<:Real}
+    return randn(length(θ))
+end
+
+function rand_momentum(h::Hamiltonian{DiagMetric{T}}, θ::AbstractVector{T}) where {T<:Real}
+    return randn(length(θ)) ./ sqrt.(h.metric.M⁻¹)
```

Comment:
If one `temp` is defined, we can do:
```julia
temp .= randn.()
temp ./= sqrt.(h.metric.M⁻¹)
```

Reply:
Are you talking about reducing memory allocation or something? And by `temp1 .= randn.()` do you mean `temp1 .= randn.(length(θ))`?

Reply:
Yes this is for reducing the memory allocations. `temp1 .= randn.()` puts a different random number in each element of `temp1`. `temp1 .= randn.(length(θ))` tries to put a random vector in each of element of `temp1` so it will error. `temp1 .= randn()` will put the *same* random number in all the elements of `temp1`. So we want the first.

Reply:
Great thanks!